### ISSUE TITLE
Numba cache files don't get invalidated with Python cache files

### ISSUE THREAD

**Issue by felixhekhorn** [numba/numba#8926]
Continuation of #8851

Dear admins, after #8851 got closed I updated the description to expose the problem again which I believe to be a numba bug. Since I didn't receive a reply so far I open this new issue - if you insist this is not a problem, please feel free to close again.

For a description of the problem see the head of #8851. In [this comment](https://github.com/numba/numba/issues/8851#issuecomment-1489960888) I phrased the problem as "the numba cache is not invalidated at the same time as the Python cache.". In [this follow-up comment](https://github.com/numba/numba/issues/8851#issuecomment-1494114896) I speculated a bit about the origin of the problem.

Thanks!

**Comment by esc** [comment]
@felixhekhorn I read through the #8851 issue and do I understand correctly, that your use-case is to store cached numba jit compiled code in a git repo? In that case perhaps AOT https://numba.readthedocs.io/en/stable/user/pycc.html is a better choice? The issue you describe seems quite tricky, and I admit, I think your best bet will be to add an entry to the FAQ.

**Comment by felixhekhorn** [comment]
Dear @esc thanks for the reply - let me try to explain a bit more ...

> The issue you describe seems quite tricky

Mmm ... I'd rather say that I do a very common task:
1. using more than 1 numba module (meaning calling inside a numba function `u` another numba function `l` from somewhere else) 
2. developing under version control
 
(the provided shell script should be sufficiently clear, no?)

> that your use-case is to store cached numba jit compiled code in a git repo

if you wish that is exactly the problem: I do _not_ want to store compiled code (`.pyc`) in the repo, but only the source code (`.py`). Storing the compiled code inside the repo would most likely solve my problem, because every time I change branch also the compiled counterpart would be exchanged, but this would completely defeat the purpose of numba: I use numba because this way I can write simple python and you guys deal with the non-trivial task of compiling to the correct architecture etc.

>  In that case perhaps AOT https://numba.readthedocs.io/en/stable/user/pycc.html is a better choice? 

Mmm, would it? keep in mind the script works just fine without the cache, so I'm quite sure the problem is there (recall that I conjecture that the numba cache is not keeping track of the call tree and so is not aware if child calls are outdated). More over I'm happy with JIT because my compilation takes quite some time, so when debugging I can sometimes skip some parts ...

> I think your best bet will be to add an entry to the FAQ.

you mean to discourse ?

**Comment by esc** [comment]
> you mean to discourse ?

Ah, no, the FAQ is here: https://numba.readthedocs.io/en/stable/user/faq.html

**Comment by esc** [comment]
> if you wish that is exactly the problem: I do _not_ want to store compiled code (`.pyc`) in the repo, but only the source code (`.py`). Storing the compiled code inside the repo would most likely solve my problem, because every time I change branch also the compiled counterpart would be exchanged, but this would completely defeat the purpose of numba: I use numba because this way I can write simple python and you guys deal with the non-trivial task of compiling to the correct architecture etc.

Does this mean, after a `git switch` you need to manually delete Numba's cache files manually for it to work?

**Comment by felixhekhorn** [comment]
> Ah, no, the FAQ is here: https://numba.readthedocs.io/en/stable/user/faq.html

ok, sorry - but there I can not write myself, no?

> Does this mean, after a `git switch` you need to manually delete Numba's cache files manually for it to work?

yes! this is exactly what the "hack" in my script is doing `# rm __pycache__/*`

**Comment by esc** [comment]
> > Ah, no, the FAQ is here: https://numba.readthedocs.io/en/stable/user/faq.html
> 
> ok, sorry - but there I can not write myself, no?

No, but you could submit a pull-request.

> > Does this mean, after a `git switch` you need to manually delete Numba's cache files manually for it to work?
> 
> yes! this is exactly what the "hack" in my script is doing `# rm __pycache__/*`

I see, I think I understand it now, thank you for the clarification.

**Comment by felixhekhorn** [comment]
I just happen to stumble about the relevant line in the docs:
https://github.com/numba/numba/blob/74cb715132519cbe5dcc4ea963c6953ef424861d/docs/source/developer/caching.rst?plain=1#L65-L66
fine, the relevant line is there - however this is not easy to find and in contrast I consider developing in several files a common use case ... so decide as you prefer ... I'd be happy if you could add support for that feature

**Comment by esc** [comment]
> I just happen to stumble about the relevant line in the docs:
> 
> https://github.com/numba/numba/blob/74cb715132519cbe5dcc4ea963c6953ef424861d/docs/source/developer/caching.rst?plain=1#L65-L66
> 
> 
> fine, the relevant line is there - however this is not easy to find and in contrast I consider developing in several files a common use case ... so decide as you prefer ... I'd be happy if you could add support for that feature

We discussed this today in the issue triage meeting and the following suggestions were mentioned:

* There are two sections in the the docs about the limitations of caching, which I think cover your use-case:
  * https://numba.readthedocs.io/en/stable/user/jit.html#cache
  * https://numba.readthedocs.io/en/stable/developer/caching.html#caching-limitations

* A "proper" fix is being attempted here:
  * https://github.com/numba/numba/pull/8396
  * Maybe @luk-f-a can comment on the above patch?

* Another idea is to "disable" the cache during development, for example by doing one or more of the following:
  * Using a git-hook to clear the cache on `git switch`
  * Setting the `NUMBA_CACHE_DIR` to a different location based on which git branch is checked out
  * Setting the `NUMBA_CACHE_DIR` to `/dev/null` 


Hope this helps?

**Comment by felixhekhorn** [comment]
I'm willing to accept and close this issue ...

>     * There are two sections in the the docs about the limitations of caching, which I think cover your use-case:
>       
>       * https://numba.readthedocs.io/en/stable/user/jit.html#cache
>       * https://numba.readthedocs.io/en/stable/developer/caching.html#caching-limitations

as said: fine, I was just not able to find on a first glance

>     * A "proper" fix is being attempted here:
>       
>       * [Extend cache invalidation to include changes in other files (take 2) #8396](https://github.com/numba/numba/pull/8396)
>       * Maybe @luk-f-a can comment on the above patch?

the description looks good - however, as this looks big (as I anticipated) also the time-scale looks big ...

>     * Another idea is to "disable" the cache during development, for example by doing one or more of the following:
>       
>       * Using a git-hook to clear the cache on `git switch`
>       * Setting the `NUMBA_CACHE_DIR` to a different location based on which git branch is checked out
>       * Setting the `NUMBA_CACHE_DIR` to `/dev/null`

this is not really an option, since we take 20min compiling the whole project - which we can not afford for a single line change .. recall this is not directly related to git, but also refers to any local change you make ...

My short-term solution is to delete always "the correct cache files" by hand. The mid-term solution will be to acknowledge that we outgrown the numba scope and use a compiled language.

In any case: thanks to everybody for your effort and support! :+1:  Numba is still a great project which simply has its limitations as everybody else.

**Comment by esc** [comment]
> In any case: thanks to everybody for your effort and support! 👍 Numba is still a great project which simply has its limitations as everybody else.

Thank you for your feedback, it sounds like you did manage to bootstrap your way into something useful! As always YMMV (your mileage may vary). Thank you for using Numba and please don't hesitate to open new issues should you encounter any snags in the future (on your current or any new projects involving Numba)! 👍
